fs2-aws-kinesis/src/main/scala/fs2/aws/kinesis/KinesisStreamBuilder.scala

```diff
@@ -0,0 +1,170 @@
+package fs2.aws.kinesis
+
+import cats.effect.kernel.Resource
+import cats.effect.std.{Dispatcher, Queue}
+import fs2.aws.kinesis.models.KinesisModels.{AppName, BufferSize, SchedulerId, StreamName}
+import fs2.{Chunk, Stream}
+import software.amazon.awssdk.services.cloudwatch.CloudWatchAsyncClient
+import software.amazon.awssdk.services.dynamodb.DynamoDbAsyncClient
+import software.amazon.awssdk.services.kinesis.KinesisAsyncClient
+import software.amazon.kinesis.common.ConfigsBuilder
+import software.amazon.kinesis.coordinator.Scheduler
+import software.amazon.kinesis.processor.{MultiStreamTracker, SingleStreamTracker, StreamTracker}
+
+abstract class KinesisStreamBuilder[F[_]] {
+
+  trait InitialPhase {
```

Comment:
this feels like a pretty restrictive builder, in that you have to call each phase, in a particular order, before `build` is available?

Reply:
correct, it has one branching, depending on what type of stream tracker you want, single stream or multistream

Reply:
Okay, but if you're forcing the caller to provide all the required `with` calls up to that branch, i.e.

``` scala
        .withAppName(appName)
        .withKinesisClient(k)
        .withDynamoDBClient(DynamoDbInterpreter[F].DynamoDbAsyncClientResource(dac))
        .withCloudWatchClient(CloudwatchInterpreter[F].CloudWatchAsyncClientResource(cac))

```

then that could all be provided in one function call saving a lot of boilerplate.  Or, put that behind a convenience function on the KinesisStreamBuilder object, e.g.

```scala
KinesisStreamBuilder.default[F](appName, kClient, dbClient, cwClient)
```

Reply:
Unless the plan is to offer alternatives in the builder (I don't see any) for those `with` calls

Reply:
also, we can do what you suggest by adding fast forward phase where all 4 params are passed and next phase would be stream tracker or scheduler, or even final phase where everything created by default (maybe need explicitly provide stream tracker as single stream and stream name to track )

Reply:
Not particularly - go for it